File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
# 프로그래머스 42627 디스크 컨트롤러 LV.3
import heapq


def solution(jobs):
    answer, now, i = 0, 0, 0
    start = -1
    wait_list = []

    while i < len(jobs):
        for j in jobs:
            if start < j[0] <= now:
                heapq.heappush(wait_list, (j[1], j[0]))

        if wait_list:
            cur = heapq.heappop(wait_list)
            start = now
            now += cur[0]
            answer += now - cur[1]
            i += 1
        else:
            now += 1

    return answer // i

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;


class Solution {
    public int solution(int[][] jobs) {
        int answer = 0;
        Arrays.sort(jobs, (o1, o2) -> o1[0] - o2[0]);
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);

        int start = -1, now = 0, i = 0;
        while (i < jobs.length) {
            for (int[] job : jobs) {
                if (start < job[0] && job[0] <= now) {
                    pq.add(job);
                }
            }

            if (!pq.isEmpty()) {
                int[] current = pq.poll();
                start = now;
                now += current[1];
                answer += now - current[0];
                i++;
            } else {
                now++;
            }
        }

        return answer / i;
    }
}